cpp/src/hash/hashing.cu

```diff
@@ -213,7 +221,168 @@ void compute_row_output_locations(size_type * __restrict__ row_partition_numbers
   }
 }
 
-template <bool has_nulls>
+/* --------------------------------------------------------------------------*/
+/** 
+ * @brief Move one column from the input table to the hashed table.
+ * 
+ * @param[in] input_buf Data buffer of the column in the input table
+ * @param[out] output_buf Preallocated data buffer of the column in the output table
+ * @param[in] num_rows The number of rows in each column
+ * @param[in] num_partitions The number of partitions to divide the rows into
+ * @param[in] row_partition_numbers Array that holds which partition each row belongs to
+ * @param[in] row_partition_offset Array that holds the offset of each row in its partition of
+ * the thread block.
+ * @param[in] block_partition_sizes Array that holds the size of each partition for each block
+ * @param[in] scanned_block_partition_sizes The scan of block_partition_sizes
+ */
+/* ----------------------------------------------------------------------------*/
+template <typename InputIter, typename DataType>
+__global__
+void move_to_output_buffer(InputIter input_iter,
+                           DataType *output_buf,
+                           const size_type num_rows,
+                           const size_type num_partitions,
+                           size_type * __restrict__ row_partition_numbers,
+                           size_type * __restrict__ row_partition_offset,
+                           size_type * __restrict__ block_partition_sizes,
+                           size_type * __restrict__ scanned_block_partition_sizes)
+{
+  extern __shared__ char shared_memory[];
+  DataType *block_output = (DataType *)shared_memory;
+  size_type *partition_offset_shared = (size_type *)(block_output + BLOCK_SIZE * ROWS_PER_THREAD);
+  size_type *partition_offset_global = (size_type *)(partition_offset_shared + num_partitions + 1);
+
+  size_type ipartition;
+
+  // Calculate the offset in shared memory of each partition in this thread block
+  if (threadIdx.x == 0) {
+    // TODO: could use a block scan instead of serialization
+    partition_offset_shared[0] = 0;
+
+    for (ipartition = 0; ipartition < num_partitions; ipartition ++) {
+      partition_offset_shared[ipartition + 1] = partition_offset_shared[ipartition]
+        + block_partition_sizes[ipartition * gridDim.x + blockIdx.x];
+    }
+  }
+
+  // Fetch the offset in the output buffer of each partition in this thread block
+  ipartition = threadIdx.x;
```

Comment:
I'd prefer this to be a separate variable since the previous value is not used after this point. This way you can pull partition above into its loop.